src/clj_irods/core.clj

```diff
@@ -0,0 +1,205 @@
+
+(defmacro maybe-icat-transaction
+  [use-icat & body]
+  `(if ~use-icat
+     (icat-direct/with-icat-transaction span-sym#
+       (do ~@body))
+     (do ~@body)))
+
+(defn make-threadpool
+  [prefix thread-count]
+  (let [counter (atom 0)]
+    (Executors/newFixedThreadPool
+      thread-count
+      (proxy [ThreadFactory] []
+        (newThread [^Runnable runnable]
+          (let [t (Thread. runnable)]
+            (.setName t (str prefix "-" (swap! counter inc)))
+            t))))))
+
+(defmacro with-irods
+  "Open connections to iRODS and/or transactions in the ICAT depending on the
+  options passed in `cfg` and what's configured.  Bind this to the symbol
+  passed in as `sym`. Recommended choice is to call it irods."
+  [cfg sym & body]
+  `(let [id# (name (gensym "with-irods-"))
+         jargon-cfg# (jargon-cfg (:jargon ~cfg))
+         use-jargon# (boolean jargon-cfg#)
+         use-icat# (have-icat)
+         jargon-pool# (or (:combined-pool ~cfg) (:jargon-pool ~cfg) (make-threadpool (str id# "-jargon") (or (:jargon-pool-size ~cfg) 1)))
+         icat-pool#   (or (:combined-pool ~cfg) (:icat-pool ~cfg) (make-threadpool (str id# "-icat") (or (:icat-pool-size ~cfg) 5)))]
+     (try+
+       (maybe-icat-transaction use-icat#
+         (maybe-jargon use-jargon# jargon-cfg# jargon#
+           (let [~sym {:jargon      jargon#
+                       :jargon-pool jargon-pool#
+                       :icat-pool   icat-pool#
+                       :has-jargon  use-jargon#
+                       :has-icat    use-icat#
+                       :cache       (atom {})}]
+             (do ~@body))))
+       (finally
+         (when-not (:retain-jargon-pool ~cfg) (.shutdown jargon-pool#))
+         (when-not (:retain-icat-pool ~cfg) (.shutdown icat-pool#))))))
+
+(defn- cached-or-get
+  [irods & get-fns]
+  (or
+    (first (filter delay?
+        (concat
+          (map #(apply (first %) true  irods (rest %)) get-fns)
+          (map #(apply (first %) false irods (rest %)) get-fns))))
+    nil))
+
+(defn- from-listing
+  [cache? irods extract-fn user zone path]
+  (let [get-from-listing (fn [listing] (first (filter #(= (:full_path %) path) (if (delay? listing) @listing listing))))]
+    (if cache?
+      (let [paged-folder-listing (icat/all-cached-listings irods user zone (ft/dirname path))
+            single-item (icat/cached-get-item irods user zone path)
+            listing-item (when (delay? paged-folder-listing) (get-from-listing (apply concat (icat/flatten-cached-listings @paged-folder-listing))))]
+        (or
+          (when listing-item (delay (extract-fn listing-item)))
+          (when single-item  (delay (extract-fn @single-item)))
+          nil))
+      (or
+        ;; better if we pass user group IDs in so we can get them from any source, rather than forcing the icat version specifically
+        (when (:has-icat irods) (delay (force (icat/user-group-ids irods user zone)) (extract-fn @(icat/get-item irods user zone path))))
```

Comment:
Will improve on this later, once there's a `clj-irods.core` function for user groups/user group ids. For now, doing this `force` means that the user group IDs will be cached, rather than ending up un-cached as part of the `get-item` call.

`get-item` itself is also a new addition. It returns exactly the same structure as a listing, but for a single path, so it's a lot like a jargon `stat` plus a couple fields (uuid and permission primarily).